### DISCUSSION TITLE
Building/Binaries for Windows

### DISCUSSION THREAD

**Discussion by PhantomzBack** [flxzt/rnote#103] [Q&A]
Hi, is there anyway at present with which I can build this for windows? Or any binaries? I found this application just today and looks really enticing, considering I wanted to make such an app too. I will try to make one when I get to college, but never the less, I wish to try this out. I am on Windows 10 currently.

**Comment by flxzt** [comment]
Hi, I don't think so atm unfortunately. There are [some instructions](https://gtk-rs.org/gtk4-rs/git/book/installation_windows.html) in the gtk-rs book, but I don't think anyone has tried them yet with rnote. If you are planning on building it on windows, please document the steps and share them here or in #46, maybe we could do a windows binary sometime in the future

**Reply by 4bitfocus** [reply]
I came here to ask this question too… kinda. I’m looking around for a UI toolkit to use with a rust project and was considering gtk4, but have heard it is challenging to build for windows.

**Reply by flxzt** [reply]
I managed to get a first build running on windows:
![rnote_on_windows](https://user-images.githubusercontent.com/19841886/167248453-4a9e238f-2800-4820-bf4e-c38766b293d8.png)


I documented the steps [here](https://github.com/flxzt/rnote/blob/main/misc/building/rnote-windows-build.md). I am getting lots of UTF-8 parsing errors though, so this is very experimental.

**Comment by Kreijstal** [comment]
I built it in windows as well..

First I already had mingw installed and some other dependencies so I didnt have to install as many dependencies, then 

`pacman -Syuu $MINGW_PACKAGE_PREFIX-rust $MINGW_PACKAGE_PREFIX-meson $MINGW_PACKAGE_PREFIX-gtk4 $MINGW_PACKAGE_PREFIX-libadwaita`
after installing all that, simply git clone deep the submodules go to the directory.
`meson setup --prefix=/usr _mesonbuild
meson compile --prefix=/usr _mesonbuild` 
To my surprise it built!
Because I was unexperienced with how this worked, I got the error
```
`$ ./rnote.exe

(rnote.exe:6200): GLib-GIO-ERROR **: 11:13:31.608: Settings schema 'com.github.flxzt.rnote' is not installed
```
So obviously I simply copied the file to `\mingw64\share\glib-2.0\schemas` and used `glib-compile-schemas` which made it work, but thats probably not the ideal way to solve it

![grafik](https://user-images.githubusercontent.com/2415206/179720723-f130ef76-536a-4a25-8673-3626da96c9e6.png)
-----
I would like to add some information, no I didn't install it on the system, because it gave me this error.
`ERROR: File 'rnote' could not be found`
I suppose because in windows, the executables are called `rnote.exe` and not `rnote`

**Reply by mjoork** [reply]
You can start executables in Windows without the extension if you have them in PATH or specify the path like `./executable`.

**Comment by flxzt** [comment]
Btw. build instructions for windows are now documented here: https://github.com/flxzt/rnote/blob/main/misc/building/rnote-windows-build.md

**Comment by danzou1ge6** [comment]
Hi, I followed the steps documented in  https://github.com/flxzt/rnote/blob/main/misc/building/rnote-windows-build.md, but the produced binary failed to execute: It exited with code 101 without outputing anything. What does this exit code mean? And how can I fix it?

My system is Windows 10.

Help appreciated.

**Reply by flxzt** [reply]
exit code 101 indicates that the program [panicked](https://doc.rust-lang.org/stable/std/macro.panic.html). Try commenting out [this line](https://github.com/flxzt/rnote/blob/main/rnote-ui/src/main.rs#L4) so that the app starts with a console and possibly run it with `RUST_BACKTRACE=full` to get a backtrace. You can either set the env var on windows through powershell or through code with something like: `env::set_var("RUST_BACKTRACE", "full");`

**Reply by danzou1ge6** [reply]
Thanks.

It turns out that rnote uses relative location `../share/rnote` to access its resources, and when it's executed in a different working directory, it can't access the resources. Now I start rnote with start menu shortcut, and it's working.

By the way, is this use of relative path of special purpose, such as compatibility under Windows?

**Reply by flxzt** [reply]
yeah for portability, there will be an installer eventually which needs that

**Comment by Kneemund** [comment]
@flxzt I created a packaging script for Rnote (based on Xournal++'s CI/CD and your building documentation) which extracts all of the runtime dependencies and resources from the MSYS2 environment. It seems to work fine - I've only had issues when using the fullscreen mode, but I think that's unrelated to the packaging. There is still quite a bit of work to do if we want to create an installer, but we could distribute "portable" Windows releases as `.zip`s for now (either manually or using GH Actions). What do you think?

EDIT: Sorry, I just realized that this is a Q&A thing. I should've probably posted that as a reply instead of a comment.

**Reply by flxzt** [reply]
Oh, that's neat. Well, I already have an inno setup installer script locally which works quite well with portable installations (as in: changing the installation path), registering the file type in the registry, working icons, .. and it would only need integration into CI/CD :)

**Reply by Kneemund** [reply]
Alright, that's even better then. Does it also extract/package the dependencies, or is that part still needed?

**Reply by flxzt** [reply]
no, not yet. it needs some manual steps before the installer script can be executed. (check matrix dm's)

**Reply by Kreijstal** [reply]
please add it to scoop :)

**Comment by Intranox** [comment]
it would be nice if this software would add some more portability by providing an executable for windows mainly but also binaries for linux. If there is a build for Windows even unstable I would be happy to try it out
